pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -276,10 +279,89 @@ func (a Adapter) waitForManifestDeployCompletion(applicationName string, gvr sch
 	}
 }
 
+// GetProtocol returns the protocol string
+func getProtocol(route routev1.Route) string {
+	if route.Spec.TLS != nil {
+		return "https"
+	}
+	return "http"
+}
+
+func pluraliseKind(gvkKind string) (kind string) {
+	// gvkKind is normally a singular noun and we need to have the kind as a plural
+	// i.e. Deploment => Deployments
+	//      Route => Routes
+	//      Ingress => Ingresses
+	kind = strings.ToLower(gvkKind + "s")
+	if strings.HasSuffix(gvkKind, "s") {
+		kind = strings.ToLower(gvkKind + "es")
+	}
+	return kind
+}
+
+func getApplicationURLFromService(client *occlient.Client, applicationName string) (fullURL string) {
+	service, err := client.GetServiceFromName(applicationName)
+	if err != nil {
+		return ""
+	}
+
+	if service.Status.LoadBalancer.Size() > 0 {
+		fullURL = fmt.Sprintf("http://%s:%d", service.Status.LoadBalancer.Ingress[0].Hostname, service.Spec.Ports[0].NodePort)
+	}
+
+	return fullURL
+}
+
+func (a Adapter) getApplicationURL(client *occlient.Client, applicationName string) (fullURL string, err error) {
+	// Need to wait for a second to give the server time to create the artifacts
+	// TODO: Replace wait with a wait for object to be created (need to determine which object!!!)
+	time.Sleep(2 * time.Second)
+
+	routeSupported, _ := client.IsRouteSupported()
+	if routeSupported {
+		labelSelector := fmt.Sprintf("%v=%v", "component", applicationName)
+		routes, err := client.ListRoutes(labelSelector)
+		if err != nil || len(routes) <= 0 {
+			// No URL found - try looking in the Service
+			fullURL = getApplicationURLFromService(client, applicationName)
+			if fullURL == "" {
+				// still no URL found - try looking for a knative Route therefore need to wait for Service and Route to be setup.
+				knGvr := schema.GroupVersionResource{Group: "serving.knative.dev", Version: "v1", Resource: "routes"}
+				knRoute, err := a.waitForManifestDeployCompletion(applicationName, knGvr, "Ready")
+				if err != nil {
+					return "", errors.Wrap(err, "error while waiting for deployment completion")
+				}
+				fullURL = knRoute.UnstructuredContent()["status"].(map[string]interface{})["url"].(string)
+			}
+		} else {
+			if len(routes) == 1 {
+				fullURL = fmt.Sprintf("%s://%s", getProtocol(routes[0]), routes[0].Spec.Host)
+			} else {
+				return "", errors.New("multiple routes found")
+			}
+		}
+	} else {
+		// Look for other resources, ie Nodeport / Ingress
+		fullURL = getApplicationURLFromService(client, applicationName)
```

Comment:
Do we still want to provide this Ingress lookup in this PR?

Reply:
Can remove it and change that back to an error, but didnt think there was any harm in leaving it in. Will remove.